### DISCUSSION TITLE
Native: Vulkan Support

### DISCUSSION THREAD

**Discussion by sjg-wdw** [maplibre/maplibre#336] [General]
## Vulkan Support

Here's some functionality we'd like to add to MapLibre. This is one proposal of several for Native and if you'd like a little context and our strategy [there's an overview](https://www.wetdogweather.com/blog/maplibre-proposals). This one is purely Native.

![Vulkan Logo](https://www.vulkan.org/user/themes/vulkan/images/logo/vulkan-logo.svg)

This is just Vulkan support.  If you're not familiar, Vulkan is an open rendering standard from the same people as recent OpenGL.  Think of it as equivalent to Metal, but for Android.  It's more than that, but that's where it falls for MapLibre.

### Drop in SDKs

After the big rendering modularization for the Metal project we should be able to drop in a new renderer.  In fact, we did drop in a new one:  Metal.  So it's possible.

In theory, we can just 'drop in' Vulkan with a bit of work that I'd break down into these categories.

#### Control Logic

This is handing the data over to Vulkan in a form it can deal with. Telling it what to render and how.  It also consists of very fiddly things like how MapLibre handles stenciling, depth, and transparency.

Our Drawable & Tweaker approach from the Metal work should move over quite well to Vulkan.

#### Shaders

In the first implementation of Metal we found we were mostly working on shaders.  That'll likely be true for Vulkan initially as well.

By the time we're done with the Metal work we'll have more complex shaders.  These should be doing more of the data driven styling themselves.  That necessitates more complexity in the programs, though handing the data to them should be the same between platforms.

#### Optimization

Every graphics SDK differs a bit in its implementation and you don't know where the bottlenecks will be.  As such, expect a long round of optimization targeting a variety of Android devices.  Where Metal tends to be predictable by device class, Vulkan takes a more *a la carte* approach, similar to OpenGL and its extensions.

You'll want to budget for a stack of Android devices for the developers. Hardware diversity is quite healthy on the Android side.

#### Conformance

MapLibre has an enormous conformance testing suite which turns up a lot of weird bugs.  The developers would spend a fair amount of time in this, as well as adding to the test coverage as problems slipped through.

#### Build System

Getting things to build for Android and possibly Windows and Linux, since they support Vulkan, is going to be hairy.  Expect that to be a bit of work for Vulkan.

### Discussion

Getting to a first version shouldn't be all that difficult.  Vulkan is low risk.  Getting to a performant and conformant version will take the rest of the time.

A single developer working for their employer could leave out half the conformance tests, or never bother to run them.  As long as it works, their employer is happy.  The MapLibre Consortium has to make it work completely.

**Comment by birkskyum** [comment]
As I understand Vulkan is a somewhat infamous for being very low level.

Might we be better off with a more high-level api similar to Metal, and get a lower maintenance burden, by using it through [wgpu](https://github.com/gfx-rs/wgpu)?

**Reply by sjg-wdw** [reply]
Linking to the [WebGPU discussion](https://github.com/maplibre/maplibre/discussions/339).

**Reply by birkskyum** [reply]
Right, sorry for repeating myself. I can't help but be slightly concerned that it could hamper Native if we add too many, or too complex, graphics APIs, due to each making it harder to add and maintain visual features.

**Reply by sjg-wdw** [reply]
Using someone else's rendering SDK potentially shifts the burden somewhere else.  Or it can if they do a decent job of it.  So it's worth asking if you can shift that burden elsewhere with something like WebGPU.

At this point in WebGPU's life cycle that's a big pile of unknowns.  So maybe?

On the subject of Vulkan, the people who want it already have OpenGL.  So why do they want Vulkan?
Performance, I'd guess.  And it's a pretty sure bet they can get it.  Vulkan has a history with game devs and is pretty well understood on its main platforms.  You can always mess up the implementation, but it won't be Vulkan's fault.

It's tempting to describe this as a choice between performance and safety.  But WebGPU isn't particularly safe at this point.  It's a new technology not very many people are familiar with and it requires a translation layer outside of the browser, as we've discussed before.  Maybe in a few years it'll be more clear.

**Reply by birkskyum** [reply]
Big pile of unknowns vs risk free. Copy that.

**Reply by sjg-wdw** [reply]
Well, nothing is risk free.  Just a known set of risks.

**Reply by sjg-wdw** [reply]
Hey, we can close this one.  Cause it works!

**Comment by birkskyum** [comment]
> Hey, we can close this one.  Cause it works!

Fantastic!  🎉
